File: Lab04/java/task-1/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.Scanner;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        int n;
        int[] v;

        private final static int MOD = 1000000007;

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextInt();
                v = new int[n + 1];
                for (int i = 1; i <= n; i++) {
                    v[i] = sc.nextInt();
                }
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(int result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", result);
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private long modpow(long base, long exp, long mod) {
            long res = 1;

            base %= mod;

            while (exp > 0) {
                if (exp % 2 == 1) {
                    res = (res * base) % mod;
                }

                exp /= 2;
                base = (base*base) % mod;
            }

            return res;
        }

        private int getResult() {
            // TODO: Aflati numarul de subsiruri (ale sirului stocat in v,
            // indexat de la 1 la n), nevide cu suma numerelor para.
            // Rezultatul se va intoarce modulo MOD (1000000007).

            long dpPar = 0;
            long dpImpar = 0;
            long elemPare = 0;

            for (int i = 0; i < n; i++) {
                if (v[i] % 2 == 1) {
                    long dpImparTemp = dpImpar;
                    dpImpar += dpPar + 1;
                    dpPar += dpImparTemp;
                } else {
                    elemPare++;
                    dpPar += dpPar + 1;
                    dpImpar += dpImpar;
                }

                dpImpar %= MOD;
                dpPar %= MOD;
            }

            long formula = modpow(2, n - (elemPare < n ? 1 : 0), MOD) - 1;

            return (int)(dpPar == formula ? dpPar : formula);
        }

        public void solve() {
            readInput();
            writeOutput(getResult());
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}

File: Lab04/cpp/task-1/main.cpp
#include <bits/stdc++.h>
using namespace std;

const int kMod = 1e9 + 7;

class Task {
 public:
    void solve() {
        read_input();
        print_output(get_result());
    }

 private:
    int n;
    vector<int> v;

    void read_input() {
        ifstream fin("in");
        fin >> n;
        v.push_back(-1); // adaugare element fictiv - indexare de la 1
        for (int i = 1, e; i <= n; i++) {
            fin >> e;
            v.push_back(e);
        }
        fin.close();
    }

    int modpow()

    int get_result() {
        /*
        Calculati numarul de subsiruri ale lui v cu suma numerelor para si
        returnati restul impartirii numarului la 10^9 + 7.
        */
        unsigned long long dpPar = 0;    // subsiruri cu suma para
        unsigned long long dpImpar = 0;  // subsiruru cu suma impara

        unsigned long long nrPar = 0;

        for (u_int i = 1; i <= n; i++) {
            if (v[i] % 2) {
                // numar impar:
                    // suma para + suma impara = suma impara
                unsigned long long dpImparTemp = dpImpar;

                    // +1 pentru subsirul cu suma impara format din acest
                    // element impar
                dpImpar += dpPar + 1;
                    // suma impara + element impar = suma para
                dpPar += dpImparTemp;
            } else {
                nrPar++;
                // numar par:
                    // suma para + suma para = suma para
                    // suma para + acest element par = suma para
                dpPar += dpPar + 1;

                    // suma impara + element par = suma impara
                    // acest element poate fi adaugat la orice "subsir impar"
                    // precedent
                dpImpar += dpImpar;
            }

            // face overflow repede si returneaza 582344007 de fiecare data
            // fa asta...
            dpImpar %= kMod;
            dpPar %= kMod;
        }

        unsigned long long formula = pow(2, n - (nrPar < n ? 1 : 0)) - 1;

        return (dpPar % kMod == formula % kMod) ? (dpPar % kMod) : (formula % kMod);
    }

    void print_output(int result) {
        ofstream fout("out");
        fout << result;
        fout.close();
    }
};

// Please always keep this simple main function!
int main() {
    // Allocate a Task object on heap in order to be able to
    // declare huge static-allocated data structures inside the class.
    Task *task = new Task();
    task->solve();
    delete task;
    return 0;
}
